File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/RequireDocHeadingsRule.java
package ai.harness.maven;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.List;
import java.util.stream.Stream;

/**
 * Rule that requires specific headings in documentation files.
 */
public class RequireDocHeadingsRule implements HarnessCheckRule {
    private static final String CATEGORY = "requireDocHeadings";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public List<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        JsonNode catNode = manifest.get(CATEGORY);
        List<String> requiredFiles = HarnessCheckHelper.extractPaths(manifest.get("requireFilesExist").get("parameters").get("paths"));
        List<String> headings = HarnessCheckHelper.extractPaths(catNode.get("parameters").get("headings"));
        JsonNode sourceFilter = catNode.get("parameters").get("sourceFilter");
        String prefix = sourceFilter.get("prefix").asText();
        String suffix = sourceFilter.get("suffix").asText();
        String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return requiredFiles.stream()
                .filter(f -> f.startsWith(prefix) && f.endsWith(suffix))
                .flatMap(f -> validateHeadings(root, f, headings, severity).stream())
                .toList();
    }

    private List<Finding> validateHeadings(Path root, String file, List<String> headings, String severity) throws MojoExecutionException {
        Path filePath = root.resolve(file);
        if (!HarnessCheckHelper.isSafeRegularFile(root, filePath)) {
            return List.of();
        }
        String text = HarnessCheckHelper.readFile(root, filePath);
        return headings.stream()
                .filter(h -> !text.contains(h))
                .map(h -> new Finding(severity, CATEGORY, "doc missing " + h + ": " + file))
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/RequireDocHeadingsRule.kt
package ai.harness.gradle

import kotlinx.serialization.json.JsonObject
import java.nio.file.Path

/**
 * Rule that requires documentation files to contain specified headings.
 */
class RequireDocHeadingsRule : HarnessCheckRule {
	override fun applies(manifest: JsonObject): Boolean {
		val category = "requireDocHeadings"
		val catObj = manifest[category]?.jsonObject ?: return false
		val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
		return enabled
	}

	override fun validate(manifest: JsonObject, root: Path, psiResults: HarnessPsiResults?): Collection<Finding> = buildSet {
		val category = "requireDocHeadings"
		val severity = HarnessCheck.Companion.severityOf(manifest, category)
		val catObj = manifest[category]?.jsonObject ?: return@buildSet
		val parametersObj = catObj["parameters"]?.jsonObject ?: return@buildSet
		val sourceCategory = HarnessCheck.Companion.stringFrom(parametersObj, "sourceFilesFromCategory")
		val sourceFiles = HarnessCheck.Companion.stringArrayFrom(manifest[sourceCategory]?.jsonObject?.get("parameters")?.jsonObject, "paths")
		val sourceFilterObj = parametersObj["sourceFilter"]?.jsonObject
		val prefix = HarnessCheck.Companion.stringFrom(sourceFilterObj, "prefix")
		val suffix = HarnessCheck.Companion.stringFrom(sourceFilterObj, "suffix")
		val headings = HarnessCheck.Companion.stringArrayFrom(parametersObj, "headings")

		val filtered = sourceFiles.filter { it.startsWith(prefix) && it.endsWith(suffix) }
		return buildSet<Finding> {
			filtered.forEach { docPath ->
				val content = HarnessCheck.Companion.readSafe(root, docPath)
				headings.forEach { heading ->
					if (!content.contains(heading)) {
						add(Finding(severity, category, "doc missing $heading: $docPath"))
					}
				}
			}
		}.toList()
	}
}
